hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/InsertIntoHoodieTableCommand.scala

```diff
@@ -95,7 +95,9 @@ object InsertIntoHoodieTableCommand extends Logging with ProvidesHoodieConfig wi
     }
     val config = buildHoodieInsertConfig(catalogTable, sparkSession, isOverWritePartition, isOverWriteTable, partitionSpec, extraOptions, staticOverwritePartitionPathOpt)
 
-    val alignedQuery = alignQueryOutput(query, catalogTable, partitionSpec, sparkSession.sessionState.conf)
+    val optimizer = sparkSession.sessionState.optimizer
+    val optimizerPlan = optimizer.execute(query)
+    val alignedQuery = alignQueryOutput(optimizerPlan, catalogTable, partitionSpec, sparkSession.sessionState.conf)
```

Comment:
is this required for case insensitivity or this is for performance optimization?

Reply:
Yeah, we need some clarification here.

Reply:
It's seems a little heavy to use optimizer just for case insensitive.
Besides, if wrap an optimize phase here, user might missed something in spark sql WEB UI.
Is there any better solution?

Reply:
This is required. I don't know how @KnightChess figured this out; I'm impressed. You can see more detail in https://github.com/apache/hudi/pull/10582, but basically the optimizer checks to make sure the output of the plan doesn't change after each optimization step. Because of the name changes, the FoldablePropagation step will fail due to pulling up some expressions into the projection where we do renaming. We now run the optimizer on the query before doing the renaming. The expressions have already been pulled up, so when that optimization step runs again later, it won't do anything.